pkg/engine/task/task_pipe.go

```diff
@@ -0,0 +1,407 @@
+	}
+
+	if pod.Spec.RestartPolicy != "" && pod.Spec.RestartPolicy != corev1.RestartPolicyOnFailure {
+		return errors.New("pipe pod RestartPolicy should be OnFailure")
+	}
+
+	sharedVolName, err := sharedVolumeName(pod)
+	if err != nil {
+		return err
+	}
+
+	mountPath, err := sharedMountPath(pod, sharedVolName)
+	if err != nil {
+		return err
+	}
+
+	// check if all referenced pipe files are children of the container mountPath
+	for _, f := range ff {
+		if !isRelative(mountPath, f.File) {
+			return fmt.Errorf("pipe file %s should be a child of %s mount path", f.File, mountPath)
+		}
+
+		fileName := path.Base(f.File)
+		// Same as K87 we use file names as ConfigMap data keys. A valid key name for a ConfigMap must consist
+		// of alphanumeric characters, '-', '_' or '.' (e.g. 'key.name',  or 'KEY_NAME',  or 'key-name', regex
+		// used for validation is '[-._a-zA-Z0-9]+')
+		if !pipeFileRe.MatchString(fileName) {
+			return fmt.Errorf("pipe file name %s should only contain alphanumeric characters, '.', '_' and '-'", fileName)
+		}
+	}
+	return nil
+}
+
+func pipePod(pod *corev1.Pod, name string) (string, error) {
+	volumeName, _ := sharedVolumeName(pod)
+	mountPath, _ := sharedMountPath(pod, volumeName)
+
+	if pod.GetAnnotations() == nil {
+		pod.SetAnnotations(make(map[string]string))
+	}
+	pod.Annotations[PipePodAnnotation] = "true"
+	pod.ObjectMeta.Name = name
+	pod.Spec.RestartPolicy = corev1.RestartPolicyOnFailure
+	pod.Spec.Containers = []corev1.Container{
+		{
+			Name:    pipePodContainerName,
+			Image:   "busybox",
+			Command: []string{"/bin/sh", "-c"},
+			Args:    []string{"sleep infinity"},
+			VolumeMounts: []corev1.VolumeMount{
+				{
+					Name:      volumeName,
+					MountPath: mountPath,
+				},
+			},
+		},
+	}
+
+	b, err := yaml.Marshal(pod)
+	if err != nil {
+		return "", fmt.Errorf("failed to create pipe task pod: %v", err)
+	}
+
+	return string(b), nil
+}
+
+func copyFiles(fs afero.Fs, ff []PipeFile, pod *corev1.Pod, ctx Context) error {
+	restCfg, err := config.GetConfig()
+	if err != nil {
+		return fatalExecutionError(fmt.Errorf("failed to fetch cluster REST config: %v", err), pipeTaskError, ctx.Meta)
+	}
+
+	var g errgroup.Group
+
+	for _, f := range ff {
+		f := f
+		log.Printf("PipeTask: %s/%s copying pipe file %s", ctx.Meta.InstanceNamespace, ctx.Meta.InstanceName, f.File)
+		g.Go(func() error {
+			// Check the size of the pipe file first. K87 has a inherent limit on the size of
+			// Secret/ConfigMap, so we avoid unnecessary copying of files that are too big by
+			// checking its size first.
+			size, err := FileSize(f.File, pod, restCfg)
+			if err != nil {
+				return fatalExecutionError(err, pipeTaskError, ctx.Meta)
+			}
+
+			if size > maxPipeFileSize {
+				return fatalExecutionError(fmt.Errorf("pipe file %s size %d exceeds maximum file size of %d bytes", f.File, size, maxPipeFileSize), pipeTaskError, ctx.Meta)
+			}
+
+			return DownloadFile(fs, f.File, pod, restCfg)
+		})
+	}
+
+	err = g.Wait()
```

Comment:
do I understand this correctly that we block the controller loop until we download the file?

Reply:
yep, this is correct.